### PR TITLE
support count(*) for Q1

### PR DESCRIPTION
**Pull request by yma11**
1. support count(*) for Q1
2. workaround for support count(*) in PlanBuilder

@Intel-bigdata/bdf-cider-engine @winningsix @jikunshang , please take a review.

### CODE REVIEW

File: velox/cider/VeloxToCiderExpr.cpp [hunk 1]
```diff
@@ -27,17 +27,19 @@ namespace {
 SQLTypeInfo getCiderType(
     const std::shared_ptr<const velox::Type>& expr_type,
     bool isNullable) {
+  // SQLTypeInfo notnull has a opposite value against isNullable
+  bool notnull = !isNullable;
```
Comment (winningsix) [inline velox/cider/VeloxToCiderExpr.cpp:31]: notNull?

File: velox/exec/tests/CiderTest.cpp [hunk 2]
```diff
@@ -92,3 +92,65 @@ TEST_F(CiderTest, compoundTest) {
...
+  std::vector<RowVectorPtr> vectors;
+  for (int32_t i = 0; i < 10; ++i) {
+    auto vector = std::dynamic_pointer_cast<RowVector>(
+        BatchMaker::createBatch(rowType_, 100, *pool_));
+    vectors.push_back(vector);
+  }
+  createDuckDbTable(vectors);
+
+  const auto plan =
+      PlanBuilder()
```
Comment (winningsix) [inline velox/exec/tests/CiderTest.cpp:138]: We should be 4 space indented or 2? In some code changes, it seems 2 space indented.

File: velox/cider/VeloxToCiderExpr.cpp [hunk 3]
```diff
@@ -305,6 +310,14 @@ std::shared_ptr<Analyzer::Expr> VeloxToCiderExprConverter::toCiderExpr(
       std::runtime_error("agg should happen on specific column.");
     }
   }
+  // count(*)
+  if (vExpr->name() == "count" && vExpr->inputs().size() == 0) {
```
Comment (winningsix) [inline velox/cider/VeloxToCiderExpr.cpp:314]: Out of scope for this pull request. can you remind me how the fallback behaviors are expected? Saying at which granularity and expected behavior accordingly?

File: velox/exec/tests/PlanBuilder.cpp [hunk 4]
```diff
@@ -235,7 +238,16 @@ PlanBuilder& PlanBuilder::aggregation(
 
     auto expr = std::dynamic_pointer_cast<const core::CallTypedExpr>(
         parseExpr(agg, planNode_->outputType()));
-    aggregateExprs.emplace_back(expr);
+    // change count_star to count for test
+    if (expr -> name() == "count_star") {
```
Comment (winningsix) [inline velox/exec/tests/PlanBuilder.cpp:242]: Hint: Let's create a method like "bool IsSupportedAggFunc(funcName)" instead of hard coding here.

**Review by winningsix — COMMENTED** [review]
Some minor comments.

File: velox/cider/VeloxToCiderExpr.cpp (hunk 3 above)
Comment (yma11) [reply velox/cider/VeloxToCiderExpr.cpp:314]: In my understanding, the expected behavior should be: 
1) throw exception(failed to translate/unsupported node) when the node can't be translated or not supported
2) catch the exception and generate nodes tree with corrected translated nodes and not translated nodes.
now we have no gap in 1) but logic of 2) is not fully implemented yet.

File: velox/exec/tests/CiderTest.cpp (hunk 2 above)
Comment (yma11) [reply velox/exec/tests/CiderTest.cpp:138]: I formatted the code with the command in VScode but sometime the indented space is 2, sometimes it's 4. Let me check about this.

File: velox/cider/VeloxToCiderExpr.cpp (hunk 3 above)
Comment (winningsix) [reply velox/cider/VeloxToCiderExpr.cpp:314]: we should have design discussion via document here. Do we have a design doc to start with?

Comment (yma11) [reply velox/cider/VeloxToCiderExpr.cpp:314]: We used to have some discussion in previous prototypes but no final decision for current design.

File: velox/exec/tests/CiderTest.cpp (hunk 2 above)
Comment (yma11) [reply velox/exec/tests/CiderTest.cpp:138]: Double checked. This line is expected.

File: velox/exec/tests/PlanBuilder.cpp (hunk 4 above)
Comment (yma11) [reply velox/exec/tests/PlanBuilder.cpp:242]: This is just a specific workaround for count(*) parse in planBuilder.

File: velox/cider/VeloxToCiderExpr.cpp [hunk 5]
```diff
@@ -280,7 +284,8 @@ std::shared_ptr<Analyzer::Expr> VeloxToCiderExprConverter::toCiderExpr(
             toCiderExpr(inputs[2], colInfo)));
   }
   // agg cases, refer RelAlgTranslator::translateAggregateRex
-  if (vExpr->name() == "sum" || vExpr->name() == "avg") {
+  if (vExpr->name() == "sum" || vExpr->name() == "avg" ||
+      (vExpr->name() == "count" && vExpr->inputs().size() == 1)) {
```
Comment (jikunshang) [inline velox/cider/VeloxToCiderExpr.cpp:288]: make `vExpr->inputs().size() == 1` as a inline function? seems hard to understand.

File: velox/cider/VeloxToCiderExpr.cpp [hunk 6]
```diff
@@ -296,15 +301,23 @@ std::shared_ptr<Analyzer::Expr> VeloxToCiderExprConverter::toCiderExpr(
...
+      //          agg_type.get_type(),
+      //          getVeloxAggType(vExpr->name(), agg_field).get_type());
       return std::make_shared<Analyzer::AggExpr>(
           agg_type, agg_kind, arg_expr, false, arg1);
     } else {
       std::runtime_error("agg should happen on specific column.");
     }
   }
+  // count(*)
+  if (vExpr->name() == "count" && vExpr->inputs().size() == 0) {
```
Comment (jikunshang) [inline velox/cider/VeloxToCiderExpr.cpp:314]: will `count(1)`  go into this branch?

File: velox/exec/tests/PlanBuilder.cpp (hunk 4 above)
Comment (jikunshang) [inline velox/exec/tests/PlanBuilder.cpp:242]: should not have space?

**Review by jikunshang — COMMENTED** [review]
overall LGTM.

File: velox/cider/VeloxToCiderExpr.cpp (hunk 3 above)
Comment (winningsix) [reply velox/cider/VeloxToCiderExpr.cpp:314]: create https://jira.devtools.intel.com/browse/POAE7-1439 for tracking this.

File: velox/cider/VeloxToCiderExpr.cpp [hunk 7]
```diff
@@ -296,15 +301,23 @@ std::shared_ptr<Analyzer::Expr> VeloxToCiderExprConverter::toCiderExpr(
...
+    SQLAgg agg_kind = getCiderAggOp(vExpr->name());
+    std::shared_ptr<Analyzer::Constant> arg1; // 2nd aggregate parameter
```
Comment (winningsix) [inline velox/cider/VeloxToCiderExpr.cpp:316]: let's use a meaningful name e.g., const_arg.

File: velox/exec/tests/PlanBuilder.cpp (hunk 4 above)
Comment (winningsix) [reply velox/exec/tests/PlanBuilder.cpp:242]: By this way, you will find you check such logic in L205 and L241. We can have  inline method like ```velox2CiderAggFunTranslate(string velox_func_name)```

File: velox/exec/tests/CiderTest.cpp (hunk 2 above)
Comment (yma11) [reply velox/exec/tests/CiderTest.cpp:138]: yes.

File: velox/cider/VeloxToCiderExpr.cpp (hunk 6 above)
Comment (yma11) [reply velox/cider/VeloxToCiderExpr.cpp:314]: I am not sure. But it's probably taken as count with constant argument. count(*) is actually taken as count with NO argument.

File: velox/cider/VeloxToCiderExpr.cpp [hunk 8]
```diff
@@ -280,7 +284,8 @@ std::shared_ptr<Analyzer::Expr> VeloxToCiderExprConverter::toCiderExpr(
             toCiderExpr(inputs[2], colInfo)));
   }
   // agg cases, refer RelAlgTranslator::translateAggregateRex
-  if (vExpr->name() == "sum" || vExpr->name() == "avg") {
+  if (vExpr->name() == "sum" || vExpr->name() == "avg" ||
```
[the remark this answers is not in the document]
Comment (yma11) [reply velox/cider/VeloxToCiderExpr.cpp]: updated. Categorize the functions for callExpression translation.

**Comment by yma11** [comment]
@winningsix @jikunshang , refine check for function translation. I will refine further on the functions support when switch to substrait.

File: velox/cider/VeloxToCiderExpr.cpp [hunk 9]
```diff
@@ -27,17 +27,19 @@ namespace {
 SQLTypeInfo getCiderType(
     const std::shared_ptr<const velox::Type>& expr_type,
     bool isNullable) {
+  // SQLTypeInfo notnull has a opposite value against isNullable
+  bool notNull = !isNullable;
```
[the remark this answers is not in the document]
Comment (winningsix) [reply velox/cider/VeloxToCiderExpr.cpp:31]: POAE7-1438 created for this. FYI

**Review by winningsix — COMMENTED** [review]
Since we're moving to Substrait, do we still need this PR merged?

**Comment by yma11** [comment]
> Since we're moving to Substrait, do we still need this PR merged?

I prefer to merge it even we will not use this repo anymore, just to save the latest code.

**Comment by winningsix** [comment]
Repo and branch are still the same. Then we need to close PR comments soon.

**Pull request closed without merging**